src/devices_models/devices/common/cost_functions.jl

```diff
@@ -232,33 +241,67 @@ function pwl_gencost_sos!(
     container::OptimizationContainer,
     spec::AddCostSpec,
     component_name::String,
-    cost_data::Vector{NTuple{2, Float64}},
     time_period::Int,
-)
-    base_power = get_base_power(container)
-    variable = get_variable(container, spec.variable_type(), spec.component_type)[
+    cost_data::Vector{NTuple{2, Float64}},
+    ::Type{T},
+) where {T <: PSY.Component}
+    return pwl_gencost_sos!(
+        container,
         component_name,
         time_period,
-    ]
+        spec.sos_status,
+        cost_data,
+        spec.variable_type,
+        T,
+    )
+end
+
+function pwl_gencost_sos!(
+    container::OptimizationContainer,
+    attributes::CostFunctionAttributes,
+    component_name::String,
+    time_period::Int,
+    cost_data::Vector{NTuple{2, Float64}},
+    ::Type{T},
+) where {T <: PSY.Component}
+    return pwl_gencost_sos!(
+        container,
+        component_name::String,
+        time_period::Int,
+        attributes.sos_status,
+        cost_data,
+        attributes.variable_type,
+        T,
+    )
+end
+
+function pwl_gencost_sos!(
+    container::OptimizationContainer,
+    component_name::String,
+    time_period::Int,
+    sos_status::SOSStatusVariable,
+    cost_data,
+    ::Type{S},
+    ::Type{T},
+) where {S <: VariableType, T <: PSY.Component}
+    base_power = get_base_power(container)
+    variable = get_variable(container, S(), T)[component_name, time_period]
     export_pwl_vars = get_export_pwl_vars(container.settings)
     @debug export_pwl_vars _group = LOG_GROUP_COST_FUNCTIONS
     total_gen_cost = JuMP.AffExpr(0.0)
 
-    if spec.sos_status == SOSStatusVariable.NO_VARIABLE
+    if sos_status == SOSStatusVariable.NO_VARIABLE
         bin = 1.0
         @debug "Using Piecewise Linear cost function but no variable/parameter ref for ON status is passed. Default status will be set to online (1.0)" _group =
             LOG_GROUP_COST_FUNCTIONS
 
-    elseif spec.sos_status == SOSStatusVariable.PARAMETER
-        param_key = encode_symbol("ON", string(spec.component_type))
+    elseif sos_status == SOSStatusVariable.PARAMETER
+        param_key = encode_symbol("ON", string(T))
```

Comment:
Is this vulnerable to a bug if T has the module name, such as `PowerSystems.ThermalStandard`?

Reply:
Should I use something like `encode_symbol("ON", string(IS.strip_module_name(T)))`

Reply:
Can you create the actual ParameterKey? Maybe something like `ParameterKey(OnStatusParameter, T)`?